### PR TITLE
Support for new interface to XYoZ quadrature

### PR DESCRIPTION
**Pull request by arporter**
This PR is for work done under #76 and changes the way we handle (existing) XYoZ quadrature.

### CODE REVIEW

**Comment by arporter** [comment]
In a bid to make this a smaller change, I've decided to stick with just supporting (the new interface to) gh_quadrature_xyoz for this issue. Once this is on master, I can then add support for the other quadrature shapes in a separate issue.

**Comment by arporter** [comment]
Documentation does not require updating as it is only the PSy layer that is changed by this issue. All files/changes are pep8/pylint clean.

**Comment by arporter** [comment]
Have gone through and checked all docstrings in modified routines (in dynamo0p3). We are now ready for first review from @rupertford

File: src/psyclone/tests/test_files/dynamo0p3/testkern_wrong_shape_qr_mod.F90 [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. Ford STFC Daresbury Lab
```
Comment (rupertford) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_wrong_shape_qr_mod.F90]: and A. R. Porter, or actually just A. R. Porter as you created this file.

**Comment by rupertford** [comment]
Back to @arporter

File: src/psyclone/dynamo0p3.py [hunk 2]
```diff
@@ -1744,10 +1816,36 @@ def __init__(self, schedule):
...
+        :param parent: the node in the f2pygen AST that will be the
+                       parent of all of the declarations (i.e. the
+                       PSy-layer subroutine)
+        :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
+        '''
+        from psyclone.f2pygen import TypeDeclGen
+        # Create a single declaration for each quadrature type
+        for shape in VALID_QUADRATURE_SHAPES:
+            if shape in self._qr_vars and self._qr_vars[shape]:
+                qr_type = shape.replace("gh_", "", 1) + "_type"
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: Good points. I've replaced with a map look-up and now use the namespace manager for the proxy variables (the others are passed to us from the Alg layer).

File: src/psyclone/dynamo0p3.py [hunk 3]
```diff
@@ -1744,10 +1816,36 @@ def __init__(self, schedule):
...
+    def declare_qr(self, parent):
+        '''
+        Create the declarations for any quadrature objects passed
+        in to an invoke. These are added as children of the supplied
+        parent node.
+
+        :param parent: the node in the f2pygen AST that will be the
+                       parent of all of the declarations (i.e. the
+                       PSy-layer subroutine)
+        :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: Done.

File: src/psyclone/dynamo0p3.py [hunk 4]
```diff
@@ -1929,6 +2019,78 @@ def initialise_basis_fns(self, parent):
...
+    def _initialise_xyz_qr(self, parent):
+        '''
+        Add in the initialisation of variables needed for XYZ
+        quadrature
+
+        :param parent: the node in the AST representing the PSy subroutine
+                       in which to insert the initialisation
+        :type parent: :py:class:``psyclone.f2pygen.SubroutineGen`
+        '''
+        if "gh_quadrature_xyz" not in self._qr_vars:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: OK

File: src/psyclone/dynamo0p3.py [hunk 5]
```diff
@@ -3136,16 +3291,25 @@ def _setup(self, ktype, module_name, args, parent):
             # dynamo 0.3 api kernels require quadrature rule arguments to be
             # passed in if one or more basis functions are used by the kernel
             # and gh_shape == "gh_quadrature_***".
-            self._qr_args = {"nh": "nqp_h",
-                             "nv": "nqp_v",
-                             "h": "wh",
-                             "v": "wv"}
+            if self._eval_shape == "gh_quadrature_xyz":
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: Done.

File: src/psyclone/dynamo0p3.py [hunk 6]
```diff
@@ -3136,16 +3291,25 @@ def _setup(self, ktype, module_name, args, parent):
...
+                self._qr_args = ["np_xyz", "weights_xyz"]
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: This is in the context of generating a kernel stub so the names are all under our control. Therefore I think it's OK.

File: src/psyclone/dynamo0p3.py [hunk 7]
```diff
@@ -3154,6 +3318,13 @@ def _setup(self, ktype, module_name, args, parent):
                 self._nodal_fspace = arg.function_space_to
             else:
                 self._nodal_fspace = arg.function_space
+        else:
+            if self._eval_shape != "":
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: Good spot, I've simplified it.

File: src/psyclone/tests/dynamo0p3_quadrature_test.py [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+        ":: basis_w2v\n"
+        "      INTEGER, intent(in) :: np_xy, np_z\n"
+        "      REAL(KIND=r_def), intent(in), dimension(np_xy) :: weights_xy\n"
+        "      REAL(KIND=r_def), intent(in), dimension(np_z) :: weights_z\n"
+        "    END SUBROUTINE dummy_code\n"
+        "  END MODULE dummy_mod")
+
+    print output
+    print str(generated_code)
+    assert str(generated_code).find(output) != -1
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_quadrature_test.py]: Done.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 9]
```diff
@@ -2053,41 +1926,46 @@ def test_op_any_space_different_space_1():  # pylint: disable=invalid-name
...
     psy = PSyFactory("dynamo0.3").create(invoke_info)
-    generated_code = psy.gen
+    generated_code = str(psy.gen)
     print generated_code
-    assert str(generated_code).find(
+
+    if utils.TEST_COMPILE:
+        # If compilation testing has been enabled (--compile flag to py.test)
+        assert utils.code_compiles("dynamo0.3", psy, tmpdir, f90, f90flags)
+    assert generated_code.find(
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_test.py]: Done.

**Comment by arporter** [comment]
Addressed all reviewer's comments and brought branch up-to-date with master.
Ready for second review from @rupertford...

**Pull request merged**